FDHA-6840 [E2E][Web Patient] NEAAR - Body Composition

alaninfo23 opened the pull request:
## Description 📝

Patient should be able to answer NEAAR-001 - Body Composition and see if label error is being displayed when user clicks on "Next" without entering a valid value on "What is your weight today?" and in "What is your body fat today".

Obs. it was also necessary to change some functions and the "Body Weight" test to work correctly.

## Demo
Body Composition:

https://github.com/alaninfo23/e2e-web-patient/assets/17017218/90906e5c-28a5-47b3-b000-5fe72423adc1

https://github.com/alaninfo23/e2e-web-patient/assets/17017218/e72ffa21-d77e-46bb-a16c-9412d35b2c58

Body Weight:

https://github.com/alaninfo23/e2e-web-patient/assets/17017218/281b08c4-a639-43e0-af0d-8688e7e85cd7

https://github.com/alaninfo23/e2e-web-patient/assets/17017218/39d358bd-4eef-482c-9078-f8b8ace46b58

## Ticket 🎟️

[FDHA-6840](https://faeththerapeutics.atlassian.net/browse/FDHA-6840)

norikoStefany-daitan commented on cypress/e2e/helpers/generalStrings.ts:
```diff
@@ -1,3 +1,3 @@
 export const PREVIOUS: string = "Previous";
 export const NEXT: string = "Next";
-export const SUBMIT_SURVEY = "Submit Survey";
+export const SUBMIT_SURVEY = "Submit Survey";
```
Please add a new line at the end of the file

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -6,16 +6,19 @@ import * as generalStrings from "../helpers/generalStrings";
-export const PROGRESS_BAR_100_IMG: string =
-  'span[role="progressbar"][aria-valuenow="100"]';
+export const VALUE_TYPE: string = "p.MuiTypography-root.css-10o8u6h";
+export const getProgressBarSelector = (progressValue: string): string => {
```
When we are creating a function that is returning a selector, we can give it the name of the selector. 
e.g.: 

> export const PROGRESS_BAR_VALUE_ID = (progressValue: string) => `span[role="progressbar"][aria-valuenow="${progressValue}"]`

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
 };
 
-export const verifyWeightScreenContent = () => {
+export const verifyScreenContent = (
```
hmm the content of this function is still specific for the Weight screen so it's better to maintain its name specific

norikoStefany-daitan commented on cypress/e2e/strings/surveyStrings.ts:
```diff
@@ -1,19 +1,27 @@
-export const PERCENT50: string = "50%";
-export const PERCENT100: string = "100%";
+export const PERCENT: string = "%";
+export const addPercentNumber = (number: string) => {
```
Same here
```suggestion
export const PERCENT_NUMBER = (number: string) => `${number}%`;
```

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
 
-export const verifyWeightScreenContent = () => {
+export const verifyScreenContent = (
+  progressBarImageSelector: string,
+  percentValue: string,
```
Instead of receiving the selectors, you could pass the values for the progressBar and percent and then inside the function here you could use the selector passing them as parameters

norikoStefany-daitan replied:
You modified only one of the parameters, progressBarImageSelector is still a selector

alaninfo23 replied:
It was just the variable name that is different, I renamed it to progressBarValue, ok?

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
+) => {
   cy.contains(BUTTON, surveyStrings.CLOSE).should("be.visible");
-  cy.contains("h4", surveyStrings.SERISD001_BODY_WEIGHT).should("be.visible");
+  cy.contains("h4", surveyTitle).should("be.visible");
```
Since its an specific function maybe we don't need to receive this as parameter

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
 };
 
-export const verifyWeightIntervalMsgError = (showMsgError: boolean) => {
+export const verifyIntervalMsgError = (
```
Now that error message can be any message I think we could give a more generic name
```suggestion
export const verifyInputMessageError = (
```

alaninfo23 committed:
FDHA-6840 [E2E][Web Patient] NEAAR - Body Composition

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
-    .should("be.visible")
-    .contains(surveyStrings.SURVEY_COMPLETED);
-  cy.wait(7000);
+export const validatetMsgOnScreen = (surveyTitle: string) => {
```
```suggestion
export const validateTitleOnScreen = (surveyTitle: string) => {
```

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
-  cy.wait(7000);
+export const validatetMsgOnScreen = (surveyTitle: string) => {
+  cy.contains("h5", surveyTitle).should("be.visible");
+};
+export const verifyBodyCompositionContentInPercent = (
```
Hmm didn't get why "InPercent" on the function name
```suggestion
};

export const verifyBodyCompositionContentInPercent = (
```

alaninfo23 replied:
This refers to the verification of body composition in percentage.

norikoStefany-daitan replied:
Got it.. but I think we should follow same pattern as the other screen and you question name to identify the function. I think on this way it gets easier to understand from which question this function is about

alaninfo23 replied:
Ok, I undertand, I renamed it to verifyBodyCompositionScreenContent, ok?

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
+  cy.contains("button", generalStrings.PREVIOUS).should("be.visible");
+};
+
+export const verifyBodyCompositionContentConfirm = (
```
Following same pattern as you used for Body Weight
```suggestion
export const verifyBodyCompositionConfirmScreen = (
```

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
   surveyName: string,
   frequency: string,
   schedule: string,
+  status: string,
```
Hmm I was thinking here, now that we are passing the schedule formatted depending on the frequency, maybe we don't need to receive `frequency` anymore. wdy?

alaninfo23 replied:
It's true, well remembered, thanks!

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -35,61 +38,129 @@ export const verifySurveyCard = (
+  cy.contains(surveyStrings.WHAT_IS_YOUR_BODY_FAT_PERCENTAGE_TODAY);
+  cy.contains(`${weightInLbs} ${surveyStrings.LBS}`);
+  cy.contains(`${bodyFatInPercent} ${surveyStrings.PERCENT}`);
+  cy.contains("button", generalStrings.PREVIOUS).should("be.visible");
```
Confirmation screen also have "Submit Survey" button. Same for Body Weight

norikoStefany-daitan replied:
Same for Body Weight

alaninfo23 replied:
True, I forgot, thanks for letting me know.

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/serabelisib/meals/bodyComposition.cy.ts:
```diff
@@ -0,0 +1,177 @@
+    adminHelper.logoutWebAdmin();
+  });
+
+  it("FDHA-6840 - Patient should be able to answer NEAAR-001 - Body Composition", () => {
```
```suggestion
  it("Patient should be able to answer NEAAR-001 - Body Composition, #3991, #6371, #6370..", () => {
```

alaninfo23 committed:
Test refactoring:

- Standardization of function names and IDs.
- change of IDs import format.
- Inversion of the test order to work around the connectUser error
- Include ticket ID in test name.

These changes aim to improve the readability and organization of the code.

alaninfo23 committed:
Test refactoring:
- Adjustment in the parameter of the verifyWeightScreenContent function.- Rename function verifyInputMessageError
- correct the name of the imports.

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -1,21 +1,41 @@
-export const PROGRESS_BAR_100_IMG: string =
-  'span[role="progressbar"][aria-valuenow="100"]';
+export const VALUE_TYPE: string = "p.MuiTypography-root.css-10o8u6h";
+export const PROGRESS_BAR_VALUE_ID = (progressValue: string): string => {
+  return `span[role="progressbar"][aria-valuenow="${progressValue}"]`;
+};
```
When we have only one statement inside {} we don't need to add them
```suggestion
export const PROGRESS_BAR_VALUE_ID = (progressValue: string): string => `span[role="progressbar"][aria-valuenow="${progressValue}"]`;
```

alaninfo23 replied:
I didn't know, good tip, thanks!

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/serabelisib/meals/bodyComposition.cy.ts:
```diff
@@ -0,0 +1,156 @@
+    logoutWebAdmin();
+  });
+
+  it("Label error should be displayed when user clicks on 'Next' without entering a valid value on What is your body fat percentage today?, #FDHA-6372", () => {
```
```suggestion
  it("Label error should be displayed when user clicks on 'Next' without entering a valid value on What is your body fat percentage today?, #6372", () => {
```

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/serabelisib/meals/bodyComposition.cy.ts:
```diff
@@ -0,0 +1,156 @@
+    validateTitleOnScreen(WHAT_IS_YOUR_WEIGHT_TODAY);
+  });
+
+  it.only("Patient should be able to answer NEAAR-001 - Body Composition, #FDHA-3991, , #FDHA-6369, #FDHA-6370, #FDHA-6371,", () => {
```
```suggestion
  it("Patient should be able to answer NEAAR-001 - Body Composition, #FDHA-3991, #FDHA-6369, #FDHA-6370, #FDHA-6371", () => {
```

alaninfo23 replied:
Sorry, I forgot to remove the only!

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/serabelisib/bodyWeight.cy.ts:
```diff
@@ -1,88 +1,113 @@
-  it("Patient should be able to answer SER-ISD-001 - Body Weight survey", () => {
-    const weightInLbs = "165.0";
-    const currentDate = moment().format("MMM D");
+  it('Label error should be displayed when user clicks on "Next" without entering a valid value weight, #FDHA-6361', () => {
```
```suggestion
  it('Label error should be displayed when user clicks on "Next" without entering a valid value weight, #6361', () => {
```

norikoStefany-daitan commented on cypress/e2e/strings/surveyStrings.ts:
```diff
@@ -1,19 +1,27 @@
-export const PERCENT50: string = "50%";
-export const PERCENT100: string = "100%";
+export const PERCENT: string = "%";
+export const PERCENT_NUMBER = (number: string) => {
+  return `${number}%`;
+};
```
When we have only one statement inside {} we don't need to add them
```suggestion
export const PERCENT_NUMBER = (number: string) => `${number}%`;
```

alaninfo23 committed:
New code improvements:
function size reduction.
rename functions with suggestive names.

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/serabelisib/meals/bodyComposition.cy.ts:
```diff
@@ -0,0 +1,156 @@
+    validateTitleOnScreen(WHAT_IS_YOUR_BODY_FAT_PERCENTAGE_TODAY);
+  });
+
+  it("FDHA-6840 - Label error should be displayed when user clicks on 'Next' without entering a valid value on What is your weight today?, #6372", () => {
```
```suggestion
  it("Label error should be displayed when user clicks on 'Next' without entering a valid value on What is your weight today?, #6372", () => {
```

alaninfo23 committed:
test name correction

alaninfo23 committed:
removing unused variables

alaninfo23 committed:
removing import variables
